## 1. Import Pandas

In [1]:
import pandas as pd

## 2. Import the Pokemon dataset (pokemon.csv)
__Note:__ If you get an error, make sure the path to the file is correct.

In [11]:
pkm = pd.read_csv("../datasets/pokemon.csv")
pkm

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

## 3. Show the first 25 entries

In [3]:
pkm.head(25)

#                       Name  Type 1  Type 2  HP  Attack  Defense  \
0    1                  Bulbasaur   Grass  Poison  45      49       49   
1    2                    Ivysaur   Grass  Poison  60      62       63   
2    3                   Venusaur   Grass  Poison  80      82       83   
3    3      VenusaurMega Venusaur   Grass  Poison  80     100      123   
4    4                 Charmander    Fire     NaN  39      52       43   
5    5                 Charmeleon    Fire     NaN  58      64       58   
6    6                  Charizard    Fire  Flying  78      84       78   
7    6  CharizardMega Charizard X    Fire  Dragon  78     130      111   
8    6  CharizardMega Charizard Y    Fire  Flying  78     104       78   
9    7                   Squirtle   Water     NaN  44      48       65   
10   8                  Wartortle   Water     NaN  59      63       80   
11   9                  Blastoise   Water     NaN  79      83      100   
12   9    BlastoiseMega Blastoise   Water     NaN  79     103      120   
13  10                   Caterpie     Bug     NaN  45      30       35   
14  11                    Metapod     Bug     NaN  50      20       55   
15  12                 Butterfree     Bug  Flying  60      45       50   
16  13                     Weedle     Bug  Poison  40      35       30   
17  14                     Kakuna     Bug  Poison  45      25       50   
18  15                   Beedrill     Bug  Poison  65      90       40   
19  15      BeedrillMega Beedrill     Bug  Poison  65     150       40   
20  16                     Pidgey  Normal  Flying  40      45       40   
21  17                  Pidgeotto  Normal  Flying  63      60       55   
22  18                    Pidgeot  Normal  Flying  83      80       75   
23  18        PidgeotMega Pidgeot  Normal  Flying  83      80       80   
24  19                    Rattata  Normal     NaN  30      56       35   

    Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0        65       65     45           1      False  
1        80       80     60           1      False  
2       100      100     80           1      False  
3       122      120     80           1      False  
4        60       50     65           1      False  
5        80       65     80           1      False  
6       109       85    100           1      False  
7       130       85    100           1      False  
8       159      115    100           1      False  
9        50       64     43           1      False  
10       65       80     58           1      False  
11       85      105     78           1      False  
12      135      115     78           1      False  
13       20       20     45           1      False  
14       25       25     30           1      False  
15       90       80     70           1      False  
16       20       20     50           1      False  
17       25       25     35           1      False  
18       45       80     75           1      False  
19       15       80    145           1      False  
20       35       35     56           1      False  
21       50       50     71           1      False  
22       70       70    101           1      False  
23      135       80    121           1      False  
24       25       35     72           1      False

## 4. Sample 15 random entries

In [4]:
pkm.sample(15)

#                     Name   Type 1    Type 2   HP  Attack  Defense  \
557  498                    Tepig     Fire       NaN   65      63       45   
245  227                 Skarmory    Steel    Flying   65      80      140   
148  137                  Porygon   Normal       NaN   65      60       70   
408  373                Salamence   Dragon    Flying   95     135       80   
51    46                    Paras      Bug     Grass   35      70       55   
84    78                 Rapidash     Fire       NaN   65     100       70   
529  477                 Dusknoir    Ghost       NaN   45     100      135   
493  445                 Garchomp   Dragon    Ground  108     130       95   
287  264                  Linoone   Normal       NaN   78      70       61   
714  647     KeldeoResolute Forme    Water  Fighting   91      72       90   
317  293                  Whismur   Normal       NaN   64      51       23   
590  531                   Audino   Normal       NaN  103      60       86   
560  501                 Oshawott    Water       NaN   55      55       45   
638  577                  Solosis  Psychic       NaN   45      30       40   
409  373  SalamenceMega Salamence   Dragon    Flying   95     145      130   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
557       45       45     45           5      False  
245       40       70     70           2      False  
148       85       75     40           1      False  
408      110       80    100           3      False  
51        45       55     25           1      False  
84        80       80    105           1      False  
529       65      135     45           4      False  
493       80       85    102           4      False  
287       50       61    100           3      False  
714      129       90    108           5      False  
317       51       23     28           3      False  
590       60       86     50           5      False  
560       63       45     45           5      False  
638      105       50     20           5      False  
409      120       90    120           3      False

## 5. What is the number of observations (rows) in the dataset?

In [5]:
pkm.shape[0]

800

## 6. What is the number of features (columns) in the dataset?

In [6]:
pkm.shape[1]


12

## 7. Print all the column names.

In [7]:
pkm.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk',
       'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

## 8. How is the dataset indexed?
__In other words:__ Print the index.

In [8]:
pkm.index

RangeIndex(start=0, stop=800, step=1)

## 9. What is the data type of each column?

In [9]:
pkm.dtypes

#              int64
Name          object
Type 1        object
Type 2        object
HP             int64
Attack         int64
Defense        int64
Sp. Atk        int64
Sp. Def        int64
Speed          int64
Generation     int64
Legendary       bool
dtype: object

## 10. Summarize the DataFrame
__Note:__ Include both numeric and non-numeric columns in the summary.

In [10]:
pkm.describe(include='all')

#       Name Type 1  Type 2          HP      Attack  \
count   800.000000        800    800     414  800.000000  800.000000   
unique         NaN        800     18      18         NaN         NaN   
top            NaN  Bulbasaur  Water  Flying         NaN         NaN   
freq           NaN          1    112      97         NaN         NaN   
mean    362.813750        NaN    NaN     NaN   69.258750   79.001250   
std     208.343798        NaN    NaN     NaN   25.534669   32.457366   
min       1.000000        NaN    NaN     NaN    1.000000    5.000000   
25%     184.750000        NaN    NaN     NaN   50.000000   55.000000   
50%     364.500000        NaN    NaN     NaN   65.000000   75.000000   
75%     539.250000        NaN    NaN     NaN   80.000000  100.000000   
max     721.000000        NaN    NaN     NaN  255.000000  190.000000   

           Defense     Sp. Atk     Sp. Def       Speed  Generation Legendary  
count   800.000000  800.000000  800.000000  800.000000   800.00000       800  
unique         NaN         NaN         NaN         NaN         NaN         2  
top            NaN         NaN         NaN         NaN         NaN     False  
freq           NaN         NaN         NaN         NaN         NaN       735  
mean     73.842500   72.820000   71.902500   68.277500     3.32375       NaN  
std      31.183501   32.722294   27.828916   29.060474     1.66129       NaN  
min       5.000000   10.000000   20.000000    5.000000     1.00000       NaN  
25%      50.000000   49.750000   50.000000   45.000000     2.00000       NaN  
50%      70.000000   65.000000   70.000000   65.000000     3.00000       NaN  
75%      90.000000   95.000000   90.000000   90.000000     5.00000       NaN  
max     230.000000  194.000000  230.000000  180.000000     6.00000       NaN

## 11. Print only the 'Name' and 'HP' columns using `.loc`

In [11]:
pkm.loc[: , ['Name', 'HP'] ]

Name  HP
0                Bulbasaur  45
1                  Ivysaur  60
2                 Venusaur  80
3    VenusaurMega Venusaur  80
4               Charmander  39
..                     ...  ..
795                Diancie  50
796    DiancieMega Diancie  50
797    HoopaHoopa Confined  80
798     HoopaHoopa Unbound  80
799              Volcanion  80

[800 rows x 2 columns]

## 12. Print only the 'Type 1' and 'Legendary' columns using `.iloc`

In [12]:
pkm.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk',
       'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

In [13]:
pkm.iloc[: , [2,-1] ]

Type 1  Legendary
0      Grass      False
1      Grass      False
2      Grass      False
3      Grass      False
4       Fire      False
..       ...        ...
795     Rock       True
796     Rock       True
797  Psychic       True
798  Psychic       True
799     Fire       True

[800 rows x 2 columns]

---
# Challenge Exercises (If You Finish Early)

### Challenge 1: Print the 'Name', 'Type 1' and 'Attack' columns of the Legendary pokemon

In [15]:
mask = pkm.loc[ : , 'Legendary' ]
mask

0      False
1      False
2      False
3      False
4      False
       ...  
795     True
796     True
797     True
798     True
799     True
Name: Legendary, Length: 800, dtype: bool

In [19]:
pkm.loc[mask, ['Name', 'Type 1', 'Attack']]

Name    Type 1  Attack
156             Articuno       Ice      85
157               Zapdos  Electric      90
158              Moltres      Fire     100
162               Mewtwo   Psychic     110
163  MewtwoMega Mewtwo X   Psychic     190
..                   ...       ...     ...
795              Diancie      Rock     100
796  DiancieMega Diancie      Rock     160
797  HoopaHoopa Confined   Psychic     110
798   HoopaHoopa Unbound   Psychic     160
799            Volcanion      Fire     110

[65 rows x 3 columns]

In [20]:
pkm.loc[ ~mask, ['Name', 'Type 1', 'Attack']]

Name  Type 1  Attack
0                Bulbasaur   Grass      49
1                  Ivysaur   Grass      62
2                 Venusaur   Grass      82
3    VenusaurMega Venusaur   Grass     100
4               Charmander    Fire      52
..                     ...     ...     ...
787    GourgeistSuper Size   Ghost     100
788               Bergmite     Ice      69
789                Avalugg     Ice     117
790                 Noibat  Flying      30
791                Noivern  Flying      70

[735 rows x 3 columns]

### Challenge 2: How many non-Legendary pokemon are there? Also list how many Pokemon per Type 1.

In [5]:
(~pkm.loc[ : , 'Legendary' ]).sum()

# Alternatively
pkm.loc[ pkm['Legendary'] == False, :].shape[0]

735

In [28]:
pkm.loc[:, "Type 1"].value_counts()

Type 1
Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: count, dtype: int64

### Challenge 3:
#### Show the `Name`, `Type 1`, and `HP` of all **Fire-type** Pokemon with `HP > 70`, sorted by `HP` descending.

In [6]:
fire = pkm.loc[ : , 'Type 1' ] == 'Fire'
high_hp = pkm.loc[:, 'HP'] > 70

pkm.loc[ fire & high_hp , ['Name', 'Type 1', 'HP']].sort_values(by='HP', ascending=False)

# Alternatively in an one-liner
# pkm.loc[(pkm.loc[:, "Type 1"] == "Fire") & (pkm.loc[: , "HP"] > 70), ["Name", "Type 1", "HP"]].sort_values(by="HP", ascending=False)

Name Type 1   HP
263                      Entei   Fire  115
559                     Emboar   Fire  110
270                      Ho-oh   Fire  106
616         DarmanitanZen Mode   Fire  105
615    DarmanitanStandard Mode   Fire  105
542                    Heatran   Fire   91
64                    Arcanine   Fire   90
158                    Moltres   Fire   90
558                    Pignite   Fire   90
736                     Pyroar   Fire   86
692                    Heatmor   Fire   85
799                  Volcanion   Fire   80
278                   Blaziken   Fire   80
279      BlazikenMega Blaziken   Fire   80
731                 Talonflame   Fire   78
6                    Charizard   Fire   78
7    CharizardMega Charizard X   Fire   78
171                 Typhlosion   Fire   78
8    CharizardMega Charizard Y   Fire   78
437                  Infernape   Fire   76
573                   Simisear   Fire   75
723                    Delphox   Fire   75
518                  Magmortar   Fire   75
43                   Ninetales   Fire   73

### Challenge 4:
#### Using only `.iloc`, print rows 100 to 110 (_remember `.iloc` is end-exclusive_) and the columns at positions `[0, 1, 5]`.

In [33]:
pkm.iloc[100:111, [0,1,5]]

#               Name  Attack
100   93            Haunter      50
101   94             Gengar      65
102   94  GengarMega Gengar      65
103   95               Onix      45
104   96            Drowzee      48
105   97              Hypno      73
106   98             Krabby     105
107   99            Kingler     130
108  100            Voltorb      30
109  101          Electrode      50
110  102          Exeggcute      40

### Challenge 5: New derived column + save CSV

Create a new column `Power` defined as `Attack + Sp. Atk + Speed`.
Then show the **top 10** Pokemon by `Power` (print `Name`, `Type 1`, `Power`), and finally, save this table to `top_power.csv` (no index).

In [34]:
pkm.loc[:, "Power"] = pkm.loc[:, 'Attack'] + pkm.loc[:, "Sp. Atk"] + pkm.loc[:, "Speed"]

Below there are some alternatives. Since we replace the original `pkm` dataframe, running more than one of these will probably fail. Pick one and start over, or save the results in new dataframe variables (i.e. `pkm1`, `pkm2`, etc.)

In [12]:
# Alternatively
# pkm.loc[:, "Power"] = pkm.loc[:, ["Attack", "Sp. Atk", "Speed"]].sum(axis=1)

In [9]:
# Alternatively!
# power = pkm['Attack'] + pkm['Sp. Atk'] + pkm['Speed']
# power.name = "Power"
# pkm = pd.concat([pkm, power], axis=1)
# pkm

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  Power  Power  
0         65       65     45           1      False    159    159  
1         80       80     60           1      False    202    202  
2        100      100     80           1      False    262    262  
3        122      120     80           1      False    302    302  
4         60       50     65           1      False    177    177  
..       ...      ...    ...         ...        ...    ...    ...  
795      100      150     50           6       True    250    250  
796      160      110    110           6       True    430    430  
797      150      130     70           6       True    330    330  
798      170      130     80           6       True    410    410  
799      130       90     70           6       True    310    310  

[800 rows x 14 columns]

In [13]:
top_p = pkm.loc[:, ["Name", "Type 1", "Power"]].sort_values(by="Power", ascending=False).head(10)
top_p.to_csv("top_power.csv", index=False)

In [15]:
less top_power.csv